perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -408,6 +408,187 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+		if (movingGroup.getName().equals(VosManager.MEMBERS_GROUP)) {
+			throw new GroupMoveNotAllowedException("It is not possible to move Members group.", movingGroup, destinationGroup);
+		}
+
+		//if destination group is null, it means group will be moved as top level group
+		if (destinationGroup != null) {
+
+			// check if both groups are from same VO
+			if (destinationGroup.getVoId() != movingGroup.getVoId()) {
+				throw new GroupMoveNotAllowedException("Groups are not from same VO. Moving group: " + movingGroup + " has VO:" + movingGroup.getVoId() + " and destination group: " + destinationGroup + " has VO:" + movingGroup.getVoId() + ".", movingGroup, destinationGroup);
+			}
+
+			// check if moving group is the same as destination group
+			if (destinationGroup.getId() == movingGroup.getId()) {
+				throw new GroupMoveNotAllowedException("Moving group: " + movingGroup + " cannot be the same as destination group: " + destinationGroup + ".", movingGroup, destinationGroup);
+			}
+
+			// check if destination group is members group
+			if (destinationGroup.getName().equals(VosManager.MEMBERS_GROUP)) {
+				throw new GroupMoveNotAllowedException("It is not possible to move group under Members group.", movingGroup, destinationGroup);
+			}
+
+			// check if moving group is already under destination group
+			if (movingGroup.getParentGroupId() != null && destinationGroup.getId() == movingGroup.getParentGroupId()) {
+				throw new GroupMoveNotAllowedException("Moving group: " + movingGroup + " is already under destination group: " + destinationGroup + " as subGroup.", movingGroup, destinationGroup);
+			}
+
+			List<Group> movingGroupAllSubGroups = getAllSubGroups(sess, movingGroup);
+
+			// check if destination group exists as subGroup in moving group subGroups
+			if (movingGroupAllSubGroups.contains(destinationGroup)) {
+				throw new GroupMoveNotAllowedException("Destination group: " + destinationGroup + " is subGroup of Moving group: " + movingGroup + ".", movingGroup, destinationGroup);
+			}
+
+			// check if this operation would create cycle
+			if (checkGroupsCycle(sess, destinationGroup.getId(), movingGroup.getId())) {
+				throw new GroupMoveNotAllowedException("There would be cycle created after moving group: " + movingGroup + " under destination group: " + destinationGroup + ".", movingGroup, destinationGroup);
+			}
+
+			List<Group> destinationGroupSubGroups = getSubGroups(sess, destinationGroup);
+
+			// check if under destination group is group with same short name as Moving group short name
+			for (Group group: destinationGroupSubGroups) {
+				if(movingGroup.getShortName().equals(group.getShortName())){
+					throw new GroupMoveNotAllowedException("Under destination group: " + destinationGroup + " is group with the same name as moving group: " + movingGroup + ".", movingGroup, destinationGroup);
+				}
+			}
+
+			//Save old parent id for moving indirect members
+			int oldParentGroupId = movingGroup.getParentGroupId();
```

Comment:
This will throw NullPointerException (I tested it) when group has no parent group, because then this attribute is null and you can't save null to int.